rust/lance/src/index/vector/builder.rs

```diff
@@ -150,78 +236,231 @@ impl<S: IvfSubIndex, Q: Quantization + Clone> IvfIndexBuilder<S, Q> {
+            lance_index::vector::ivf::new_ivf_transformer_with_quantizer(
+                ivf.centroids.clone().unwrap(),
+                self.distance_type,
+                &self.column,
+                quantizer.into(),
+                Some(0..ivf.num_partitions() as u32),
+            )?,
+        );
         let mut transformed_stream = Box::pin(
-            stream
-                .map(move |batch| {
-                    let ivf_transformer = transformer.clone();
-                    tokio::spawn(async move { ivf_transformer.transform(&batch?) })
-                })
-                .buffered(num_cpus::get())
-                .map(|x| x.unwrap())
-                .peekable(),
+            data.map(move |batch| {
+                let ivf_transformer = transformer.clone();
+                tokio::spawn(async move { ivf_transformer.transform(&batch?) })
+            })
+            .buffered(num_cpus::get())
+            .map(|x| x.unwrap())
+            .peekable(),
         );
 
         let batch = transformed_stream.as_mut().peek().await;
         let schema = match batch {
             Some(Ok(b)) => b.schema(),
             Some(Err(e)) => panic!("do this better: error reading first batch: {:?}", e),
-            None => panic!("no data"),
+            None => {
+                log::info!("no data to shuffle");
+                self.shuffle_reader = Some(Box::new(IvfShufflerReader::new(
+                    self.dataset.object_store().clone(),
+                    self.temp_dir.clone(),
+                    vec![0; ivf.num_partitions()],
+                )));
+                return Ok(self);
+            }
         };
 
-        self.shuffler
-            .shuffle(Box::new(RecordBatchStreamAdapter::new(
-                schema,
-                transformed_stream,
-            )))
-            .await
+        self.shuffle_reader = Some(
+            self.shuffler
+                .shuffle(Box::new(RecordBatchStreamAdapter::new(
+                    schema,
+                    transformed_stream,
+                )))
+                .await?,
+        );
+
+        Ok(self)
+    }
+
+    async fn build_partitions(&mut self) -> Result<&mut Self> {
+        let ivf = self.ivf.as_ref().ok_or(Error::invalid_input(
+            "IVF not set before building partitions",
+            location!(),
+        ))?;
+
+        let reader = self.shuffle_reader.as_ref().ok_or(Error::invalid_input(
+            "shuffle reader not set before building partitions",
+            location!(),
+        ))?;
+
+        let partition_build_order = (0..ivf.num_partitions())
+            .map(|partition_id| reader.partiton_size(partition_id))
+            .collect::<Result<Vec<_>>>()?
+            // sort by partition size in descending order
+            .into_iter()
+            .enumerate()
+            .map(|(idx, x)| (x, idx))
+            .sorted()
+            .rev()
+            .map(|(_, idx)| idx)
+            .collect::<Vec<_>>();
```

Comment:
How expensive is `reader.partition_size()`? I feel like this could simply be:
```suggestion
        let mut partition_build_order = (0..ivf.num_partitions()).collect::<Vec<_>>();
        partition_build_order.sort_by_key(|part_id| -reader.partition_size(part_id));
```

Reply:
this can't be done because:
- `partition_size` returns `Result<usize>` 
- can'y apply "-" on usize
but i simplify this a little